### PR TITLE
docs and public directories

### PR DESCRIPTION
**Pull request by sjoudaki**
Generated using the commands (added here for completeness; after executing "python setup.py develop"):

i) sphinx-apidoc -feo docs rail
ii) sphinx-build docs public/

We obtain two sphinx warnings:

i) WARNING: autodoc: failed to import module 'flexzboost' from module 'rail.estimation.algos'; the following exception was raised: No module named 'flexcode'
ii) checking consistency... /mnt/zfsusers/sjoudaki/railstuff/RAIL/docs/modules.rst: WARNING: document isn't included in any toctree

It seems we might be ok with these as flexzboost is not necessarily part of the fiducial installation of RAIL?

### CODE REVIEW

**Comment by sjoudaki** [comment]
The two files used to generate the rest are docs/conf.py and docs/Makefile.

**Comment by sjoudaki** [comment]
(conf_testing.py can be removed)

File: docs/conf_testing.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+master_doc = 'index'
+
+# General information about the project.
+project = 'RAIL'
+copyright = '2020-2021, LSST DESC RAIL Contributors'
+author = 'LSST DESC RAIL Contributors'
+language = 'en'
+
+# Files to ignore when looking for source files
+exclude_patterns = []
```
Comment (johannct) [inline docs/conf_testing.py]: Here you can exclude the rst files provoking the warnings I think : https://stackoverflow.com/questions/43801615/python-sphinx-exclude-patterns

File: docs/index.rst [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+   :caption: Examples
+
+.. toctree::
+   :maxdepth: 1
+   :caption: Usage Demos
+
+.. toctree::
+   :maxdepth: 1
+   :caption: Reference
+   api
```
Comment (johannct) [inline docs/index.rst]: The index page does not seem to be properly rendered. Did I do smthg wrong?
Screenshot from 2020-12-15 12-37-11

Comment (johannct) [reply docs/index.rst]: when I click on "view page source" it points to file:///home/cohen/lsst/codes/pipelines/photoz/RAIL/public/_sources/index.rst.txt.... I think the problem is that the index should actually be file:///home/cohen/lsst/codes/pipelines/photoz/RAIL/public/rail.html but the fact that there is a top RAIL repository with the package rail within may trouble sphinx as it is currently configured.

File: docs/rail.estimation.estimator.rst [hunk 3]
```diff
@@ -0,0 +1,7 @@
+rail.estimation.estimator module
+================================
+
+.. automodule:: rail.estimation.estimator
+   :members:
+   :undoc-members:
+   :show-inheritance:
```
Comment (johannct) [inline docs/rail.estimation.estimator.rst]: Do we want to fix the doc together with this pull request? I see that sphinx does not seem to like blank line; the `estimate` docstring in Estimator class is 
```
"""
        The main run method for the photo-z, should be implemented in
        the specific subclass

        Input:
        ------
        data:
          dictionary of all input data

        Returns:
        --------
        pz_dict:
          dictionary of output photo-z params, must include zmode and
          pdf note: zgrid will still be a class variable for now

        should create photo-z estimates with set names, TBD
        for demo will just be `z_mode`
        """
```
and is rendered only as `Prior settings and/or training algorithm for the individual photo-z method, should be implemented in the subclass`

**Review by johannct — CHANGES_REQUESTED** [review]
I think the `public` dir needs to be removed from git and the pull request. I do not think that we want to ship the result of the doc build with git. If we intend to deploy the doc publicly, then we will need a github actions that does the build and deploy to a public space.
On the other hand, we should probably have a ` RAIL/.github/workflows/doc.yml` action to trigger doc building on releases....

**Comment by sjoudaki** [comment]
Here, running:

(base) IT222222:RAIL sjoudaki$ sphinx-apidoc -feo docs .
Creating file docs/setup.rst.
Creating file docs/rail.rst.
Creating file docs/rail.creation.rst.
Creating file docs/rail.estimation.estimator.rst.
Creating file docs/rail.estimation.utils.rst.
Creating file docs/rail.estimation.rst.
Creating file docs/rail.estimation.algos.flexzboost.rst.
Creating file docs/rail.estimation.algos.randomPZ.rst.
Creating file docs/rail.estimation.algos.sklearn_nn.rst.
Creating file docs/rail.estimation.algos.trainZ.rst.
Creating file docs/rail.estimation.algos.rst.
Creating file docs/rail.evaluation.rst.
Creating file docs/modules.rst.


(base) IT222222:RAIL sjoudaki$ sphinx-build docs public/
Running Sphinx v1.7.4
loading translations [en]... done
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: 0 added, 3 changed, 0 removed
estimator flexzboost not installedimation.algos.flexzboost                                                                                                                                             
/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Found classifier randomPZ
Found classifier simpleNN
Found classifier trainZ
reading sources... [100%] setup                                                                                                                                                                        
WARNING: autodoc: failed to import module 'rail.estimation.algos.flexzboost'; the following exception was raised:
No module named 'flexcode'
WARNING: autodoc: failed to import module 'setup'; the module executes module level statement and it might call sys.exit().
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] setup                                                                                                                                                                         
generating indices... genindex py-modindex
highlighting module code... [100%] rail.estimation.utils                                                                                                                                               
writing additional pages... search
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 2 warnings.

The HTML pages are in public.

---

As one can see, runs fine except two warnings, one on flexzboost and one on setup.py.

File: docs/conf_testing.py (hunk 1 above)
Comment (sjoudaki) [reply docs/conf_testing.py]: Sounds good, thanks. In the latest conf.py we have a few items there. 

exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']

Could add more if we wish, but I think it's better to receive warnings on e.g. flexzboost so one is kept aware of it.

File: docs/index.rst (hunk 2 above)
Comment (sjoudaki) [reply docs/index.rst]: The index file should be working now, looks like this at the moment.

Screen Shot 2021-01-28 at 4 50 21 PM

File: docs/rail.estimation.estimator.rst (hunk 3 above)
Comment (sjoudaki) [reply docs/rail.estimation.estimator.rst]: I think sphinx is generating this exactly as it has been written now, please let me know your thoughts?

**Review by johannct — COMMENTED** [review]
index.rst is not supposed to  be regenerated?

**Comment by johannct** [comment]
> index.rst is not supposed to be regenerated?

no it does not appear to be generated....

**Comment by sjoudaki** [comment]
> > index.rst is not supposed to be regenerated?
> 
> no it does not appear to be generated....

yes, that's right, need to keep this file.

**Comment by sjoudaki** [comment]
I merged the issue/34/autodoc-debug branch, resolving the MR conflicts in the process so that the version in this branch continues to be functional but minimal. 

@aimalz the issue with the sidebar toggle should also be resolved now. Still need to decide whether we want the creation, estimation, and evaluation subpackages listed on the index page (and if so, resolve why only the contents of creation are shown, and same with its appearance in the sidebar).

**Review by johannct — APPROVED** [review]

**Comment by sjoudaki** [comment]
Just adding a comment here that this is now being merged thanks to work by @johannct which led to updates to the files. (and thanks also to @aimalz for earlier updates from his autodoc-debug branch)

**Comment by sjoudaki** [comment]
This is merged now. Thank you so much again @aimalz and @johannct for all of your efforts on this PR! Minor note for the future: if we want to revive the "Edit on GitHub" button (which we decided to remove given Slack discussion), then this is achieved by simply removing the breadcrumbs.html file in the _templates directory.

**Pull request merged**